### PR TITLE
Add UDT support to System.Data.SqlClient.

### PR DESCRIPTION
**Pull request by corivera**

### CODE REVIEW

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlFunctionAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            _isDeterministic = false;
+            _dataAccess = DataAccessKind.None;
+            _systemDataAccess = SystemDataAccessKind.None;
+            _isPrecise = false;
+            _name = null;
+            _tableDefinition = null;
+            _fillRowMethodName = null;
+        }
+
+        public bool IsDeterministic
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlFunctionAttribute.cs]: Nit: These properties can be written in Short hand using `=>`  
```
public bool IsDeterministic 
{
   get => _isDeterministic;
   set => _isDeterministic = value;
}
```

Applies to all properties.

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs [hunk 2]
```diff
@@ -264,6 +285,13 @@ public SqlCompareOptions CompareOptions
             }
         }
 
+        public DbType DbType
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: Nit: 
Use expression body definition. 
```
public DbType DbType => sxm_rgSqlDbTypeToDbType[(int)_sqlDbType];
```

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs [hunk 3]
```diff
@@ -345,17 +373,40 @@ public SqlDbType SqlDbType
             }
         }
 
+        public Type Type
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: Nit: Use Expression body definition

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs [hunk 4]
```diff
@@ -345,17 +373,40 @@ public SqlDbType SqlDbType
...
+                    return UdtTypeName;
+                }
+                else
                 {
                     return sxm_rgDefaults[(int)SqlDbType].Name;
                 }
             }
         }
 
+        internal string ServerTypeName
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: Nit: Use Expression body definition

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs [hunk 5]
```diff
@@ -961,7 +1052,7 @@ public SqlBinary Adjust(SqlBinary value)
             {
                 byte[] rgbValue = value.Value;
                 byte[] rgbNewValue = new byte[MaxLength];
-                Buffer.BlockCopy(rgbValue, 0, rgbNewValue, 0, (int)MaxLength);
+                Array.Copy(rgbValue, rgbNewValue, (int)MaxLength);
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: Why did you change this to Array.Copy here and the rest of the occurrences in this file? 
Isn't `Buffer.BlockCopy` more performant for byte array copy?

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMethodAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+        public SqlMethodAttribute()
+        {
+            // default values
+            _isCalledOnNullInputs = true;
+            _isMutator = false;
+            _shouldInvokeIfReceiverIsNull = false;
+        }
+
+        public bool OnNullCall
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/Sql/SqlMethodAttribute.cs]: Nit: expression body definition can be used in this class

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs [hunk 7]
```diff
@@ -161,6 +165,13 @@ internal SqlConnectionString.TypeSystem TypeSystem
             }
         }
 
+        internal Version TypeSystemAssemblyVersion
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlConnection.cs]: Nit: expression body definition can be used here

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlParameter.cs [hunk 8]
```diff
@@ -1793,87 +1890,161 @@ private int ValueSize(object value)
...
         // of this and use a simple regex to do the parsing
-        internal static string[] ParseTypeName(string typeName)
+        internal static string[] ParseTypeName(string typeName, bool isUdtTypeName)
         {
             Debug.Assert(null != typeName, "null typename passed to ParseTypeName");
 
             try
             {
                 string errorMsg;
+                if (isUdtTypeName)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlParameter.cs]: Nit: `errorMsg = isUdtTypeName ? SR.SQL_UDTTypeName : SR.SQL_TypeName;`

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 9]
```diff
@@ -6837,9 +6872,13 @@ internal Task TdsExecuteSQLBatch(string text, int timeout, SqlNotificationReques
                         _connHandler.CheckEnlistedTransactionBinding();
 
                         stateObj.SetTimeoutSeconds(timeout);
+
                         stateObj.SniContext = SniContext.Snix_Execute;
 
-                        WriteRPCBatchHeaders(stateObj, notificationRequest);
+                        if (_isYukon)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: _isYukon is 2005 and above right? Do we want to add this condition here? .Net Core support is for all supported Sql Server which are 2005 and above.

File: src/System.Data.SqlClient/tests/ManualTests/SQL/UdtTest/SqlServerTypesTest.cs [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+                        }
+                        bytes = buffer.ToArray();
+                        Assert.Equal("5ade", ToHexString(bytes));
+                    }
+                }
+            }
+        }
+
+        [CheckConnStrSetupFact]
+        public static void TestUdtSchemaMetadata()
```
Comment (saurabh500) [inline src/System.Data.SqlClient/tests/ManualTests/SQL/UdtTest/SqlServerTypesTest.cs]: Can we have a test for Udt as a parameter ? This can be handled in a separate PR though.

Comment (saurabh500) [reply src/System.Data.SqlClient/tests/ManualTests/SQL/UdtTest/SqlServerTypesTest.cs]: Ignore this. Github review hadn't expanded the Diffs for a few tests. I see that there are tests for Udts in SqlParameter

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 9 above)
Comment (corivera) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: We've left it out before. I'll remove these Yukon checks.

File: src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs (hunk 5 above)
Comment (corivera) [reply src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: The original framework implementation used Array.Copy. Should I switch it back to Buffer.BlockCopy?

Comment (saurabh500) [reply src/System.Data.SqlClient/src/System/Data/Sql/SqlMetaData.cs]: Yes. Please switch it back.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 9 above)
Comment (corivera) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Looks like lots of other locations in this file also check for Yukon. I'll remove them in a separate PR.

Comment (saurabh500) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Sounds good

File: src/System.Data.SqlClient/dir.props [hunk 11]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="..\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.3.0.0</AssemblyVersion>
+    <AssemblyVersion>4.3.1.0</AssemblyVersion>
```
Comment (weshaggard) [inline src/System.Data.SqlClient/dir.props]: For new APIs we generally bump the minor version.

File: src/System.Data.SqlClient/ref/System.Data.SqlClient.csproj [hunk 12]
```diff
@@ -6,7 +6,7 @@
     <IsPartialFacadeAssembly Condition="'$(TargetGroup)' == 'netfx'">true</IsPartialFacadeAssembly>
     <!-- Must match version supported by frameworks which support 4.3.* inbox.
          Can be removed when API is added and this assembly is versioned to 4.4.* -->
-    <AssemblyVersion Condition="'$(TargetsNetFx)' != 'true'">4.3.0.0</AssemblyVersion>
+    <AssemblyVersion Condition="'$(TargetsNetFx)' != 'true'">4.3.1.0</AssemblyVersion>
```
[the remark this answers is not in the document]
Comment (weshaggard) [reply src/System.Data.SqlClient/ref/System.Data.SqlClient.csproj]: See https://github.com/dotnet/corefx/issues/23743 for more details.
